apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc.rs

```diff
@@ -0,0 +1,428 @@
+// gas per block, otherwise a proposal will be rejected!
+
+use std::marker::PhantomData;
+
+use num_rational::Ratio;
+
+use crate::facade::tendermint_proto::abci::RequestPrepareProposal;
+
+/// All status responses from trying to allocate block space for a tx.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum AllocStatus {
+    /// The transaction is able to be included in the current block.
+    Accepted,
+    /// The transaction can only be included in an upcoming block.
+    Rejected { tx_len: u64, space_left: u64 },
+    /// The transaction would overflow the allotted bin space,
+    /// therefore it needs to be handled separately.
+    OverflowsBin { tx_len: u64, bin_size: u64 },
+}
+
+/// Allotted space for a batch of transactions in some proposed block,
+/// measured in bytes.
+///
+/// We keep track of the current space utilized by:
+///
+///   - Protocol transactions.
+///   - DKG decrypted transactions.
+///   - DKG encrypted transactions.
+#[derive(Debug, Default)]
+pub struct BlockSpaceAllocator<State> {
+    /// The current state of the [`BlockSpaceAllocator`] state machine.
+    _state: PhantomData<*const State>,
+    /// The total space Tendermint has allotted to the
+    /// application for the current block height.
+    block: TxBin,
+    /// The current space utilized by protocol transactions.
+    protocol_txs: TxBin,
+    /// The current space utilized by DKG encrypted transactions.
+    encrypted_txs: TxBin,
+    /// The current space utilized by DKG decrypted transactions.
+    decrypted_txs: TxBin,
+}
+
+impl From<&RequestPrepareProposal>
+    for BlockSpaceAllocator<states::BuildingDecryptedTxBatch>
+{
+    #[inline]
+    fn from(req: &RequestPrepareProposal) -> Self {
+        let tendermint_max_block_space_in_bytes = req.max_tx_bytes as u64;
+        Self::init(tendermint_max_block_space_in_bytes)
+    }
+}
+
+impl BlockSpaceAllocator<states::BuildingDecryptedTxBatch> {
+    /// Construct a new [`BlockSpaceAllocator`], with an upper bound
+    /// on the max size of all txs in a block defined by Tendermint.
+    #[inline]
+    pub fn init(tendermint_max_block_space_in_bytes: u64) -> Self {
+        let max = tendermint_max_block_space_in_bytes;
+        Self {
+            _state: PhantomData,
+            block: TxBin::init(max),
+            protocol_txs: TxBin::default(),
+            encrypted_txs: TxBin::default(),
+            // decrypted txs can use as much space as needed; in practice,
+            // we'll only need, at most, the amount of space reserved for
+            // encrypted txs at the prev block height
+            decrypted_txs: TxBin::init(max),
+        }
+    }
+}
+
+impl<State> BlockSpaceAllocator<State> {
+    /// Return uninitialized space in tx bins, resulting from ratio conversions.
+    ///
+    /// This method should not be used outside of [`BlockSpaceAllocator`]
+    /// instance construction or unit testing.
+    #[inline]
+    fn uninitialized_space_in_bytes(&self) -> u64 {
```

Comment:
What about `free_space_in_bytes` or `remaining_space_in_bytes` instead of `uninitialized_space_in_bytes`? I'm not clear on the differences between these terms: allotted, allocating, initialized - maybe there is some overlap between them

Reply:
let's say `B` is the total space Tendermint has given the application for a certain block proposal. an individual allotment `A` is a fraction of `B`, which is the total space ceded to some tx bin. for a given allocation `a`, `a <= A`, for some tx bin `T`